File: py/2021/02/main.py
import argparse
import os
import sys


def get_puzzle_input():
    """
    Handles input file retrieval with the option to override the default path.
    """
    # Get the day and year directory this script is stored in
    day = os.path.basename(os.path.dirname(os.path.abspath(__file__)))
    year = os.path.basename(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

    # Build the default input path
    root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../.."))
    default_input_path = os.path.join(root_dir, "inputs", year, day, "input.txt")

    parser = argparse.ArgumentParser(description="Advent of Code Solution")
    parser.add_argument(
        "-i", "--input",
        help="Specify a different puzzle input file path",
        default=default_input_path
    )
    args = parser.parse_args()

    input_path = args.input

    if not os.path.exists(input_path):
        print(f"Error: Input file '{input_path}' does not exist.", file=sys.stderr)
        sys.exit(66)

    return input_path


def main():
    input_path = get_puzzle_input()

    iInput = []
    with open(input_path, "r") as f:
        for line in f:
            iInput.append(line)

    # Part 1
    horizontal = 0
    depth = 0

    for instruction in iInput:
        temp = instruction.split()
        direction = temp[0]
        value = int(temp[1])
        if direction == "forward":
            horizontal += value
        elif direction == "down":
            depth += value
        elif direction == "up":
            depth -= value

    print(horizontal * depth)

    # Part 2
    horizontal = 0
    depth = 0
    aim = 0

    for instruction in iInput:
        temp = instruction.split()
        direction = temp[0]
        value = int(temp[1])
        if direction == "forward":
            horizontal += value
            depth += aim * value
        elif direction == "down":
            aim += value
        elif direction == "up":
            aim -= value

    print(horizontal * depth)


if __name__ == "__main__":
    main()

File: cs/2021/02/main.cs

using System.IO;

namespace Solver
{

    public enum Direction
    {
        UP,
        DOWN,
        FORWARD
    }

    public struct Instruction
    {
        public Direction direction;
        public int units;
    }

    public struct Position
    {
        public int horizontal;
        public int depth;
        public int aim;

        public void RunInstruction(Instruction instruction)
        {
            switch (instruction.direction)
            {
                case Direction.UP:
                    {
                        depth -= instruction.units;
                        break;
                    }
                case Direction.DOWN:
                    {
                        depth += instruction.units;
                        break;
                    }
                case Direction.FORWARD:
                    {
                        horizontal += instruction.units;
                        break;
                    }
            }
        }

        public void RunWithAim(Instruction instruction)
        {
            switch (instruction.direction)
            {
                case Direction.UP:
                    {
                        aim -= instruction.units;
                        break;
                    }
                case Direction.DOWN:
                    {
                        aim += instruction.units;
                        break;
                    }
                case Direction.FORWARD:
                    {
                        horizontal += instruction.units;
                        depth += aim * instruction.units;
                        break;
                    }
            }
        }
    }

    public class Solver
    {
        public Solver() { }

        public static List<Instruction> GenerateInput(String filePath)
        {
            List<Instruction> output = new();

            using (StreamReader reader = new(filePath))
            {
                string? line;
                while ((line = reader.ReadLine()) != null)
                {
                    Instruction instruction = new();
                    string[] fields = line.Split(" ");

                    Enum.TryParse(fields[0].ToUpper(), out instruction.direction);
                    int.TryParse(fields[1], out instruction.units);

                    output.Add(instruction);
                }
            }

            return output;
        }

        private static string GetInputFilePath(string[] args)
        {
            for (int i = 0; i < args.Length; i++)
            {
                if (args[i] == "-i" && i + 1 < args.Length)
                {
                    string inputPath = args[i + 1];
                    if (File.Exists(inputPath))
                    {
                        return inputPath;
                    }
                    else
                    {
                        throw new FileNotFoundException($"The specified file does not exist: {inputPath}");
                    }
                }
            }

            string defaultPath = Path.Combine("..", "..", "..", "inputs", "2021", "02", "input.txt");

            if (!File.Exists(defaultPath))
            {
                throw new FileNotFoundException($"The default input file was not found at {defaultPath}");
            }

            return defaultPath;
        }

        static void Main(string[] args)
        {
            string inputFilePath = GetInputFilePath(args);
            List<Instruction> inputData = GenerateInput(inputFilePath);

            Position part1Position = new();
            Position part2Position = new();

            foreach (Instruction instruction in inputData)
            {
                part1Position.RunInstruction(instruction);
                part2Position.RunWithAim(instruction);
            }

            Console.WriteLine($"Part One: {part1Position.horizontal * part1Position.depth}");
            Console.WriteLine($"Part Two: {part2Position.horizontal * part2Position.depth}");
        }

    }

}